### ISSUE TITLE
ndcg calculation is unstable with respect to the order of inputs

### ISSUE THREAD

**Issue by ebernhardson** [dmlc/xgboost#3050]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System: Debian jessie

Compiler: openjdk 1.8,0_131, gcc 4.9.2-10

Package used (python/R/jvm/C++): xgboost4j, but problem looks to be in the C++ side

`xgboost` version used: xgboost4j with master @ 3f3f54b, python with 0.7.post3 from pypi

## Steps to reproduce

```
import ml.dmlc.xgboost4j.LabeledPoint
import ml.dmlc.xgboost4j.scala.{DMatrix, XGBoost}
val dmat = new DMatrix(Iterator(LabeledPoint(1, null, Array(1,2)), LabeledPoint(2, null, Array(1, 2))))
val booster = ml.dmlc.xgboost4j.scala.XGBoost.train(dmat, Map("objective" -> "rank:ndcg", "eval_metric" -> "ndcg@1"), 1)
booster.predict(dmat)
// Array[Array[Float]] = Array(Array(0.5), Array(0.5))

booster.evalSet(Array(dmat), Array("test"), 0)
// String = [0]    test-ndcg@1:0.333333

val dmat2 = new DMatrix(Iterator(LabeledPoint(2, null, Array(1,2)), LabeledPoint(1, null, Array(1, 2))))
booster.evalSet(Array(dmat2), Array("test"), 0)
// String = [0]    test-ndcg@1:1.000000
```

Looks like the same in python:
```
import xgboost
import numpy as np

dmat = xgboost.DMatrix(np.array([[1,2], [2,1]]), label=np.array([1,2]))                                                                                                             
dmat.set_group(np.array([2]))
model = xgboost.train(params={"objective": "rank:ndcg", "eval_metric": "ndcg@1"}, dtrain=dmat, num_boost_round=1)                                                                   
model.predict(dmat)
# array([0.5, 0.5], dtype=float32)
model.eval(dmat)
# u'[0]\teval-ndcg@1:0.333333'

dmat2 = xgboost.DMatrix(np.array([[1,2], [2,1]]), label=np.array([2,1]))
dmat2.set_group(np.array([2]))
model.eval(dmat2)
# u'[0]\teval-ndcg@1:1.000000'

```
It seems when the ndcg calculation sorts the predictions the sort of equal valued predictions is unstable. Having equal predictions should be incredibly rare in practice, but it seems in the case predictions are equal it should then sort them by label (either up or down, doesn't matter, as long as it's stable).
